Simplify requirement for what we consider a setter method in ClassAndPropertiesSettersJavadocScraper

Noticed at https://github.com/checkstyle/checkstyle/pull/13678#issuecomment-1702760681
```
[INFO] Running com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.181 s <<< FAILURE! -- in com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
[ERROR] com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest -- Time elapsed: 1.181 s <<< ERROR!
java.lang.NullPointerException
```

The problem was noticed with MatchXpathCheck's `setQuery()`. The current implementation of `isSetterMethod()` returns `false` for that setter. In particular at:
```java
                // Maximum nodes allowed in a body of setter
                final int setterBodySize = 3;
                if (slist != null && slist.getChildCount() == setterBodySize) {
                    final DetailAST expr = slist.getFirstChild();
                    setterMethod = expr.getFirstChild().getType() == TokenTypes.ASSIGN;
                }
```

The `slist` has 4 children:
![Screenshot from 2023-09-01 15-52-07](https://github.com/checkstyle/checkstyle/assets/23459549/3cfd7bd0-1634-4f0b-a38b-aea6abd2cee4)

--- a/config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml
+++ b/config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml
@@ -6228,17 +6228,6 @@
 
   <checkerFrameworkError unstable="false">
     <fileName>src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java</fileName>
-    <specifier>argument</specifier>
-    <message>incompatible argument for parameter value of putIfAbsent.</message>
-    <lineContent>javadocs.putIfAbsent(property, SUPER_CLASS_PROPERTIES_JAVADOCS.get(property));</lineContent>
-    <details>
-      found   : @Initialized @Nullable DetailNode
-      required: @Initialized @NonNull DetailNode
-    </details>
-  </checkerFrameworkError>
-
-  <checkerFrameworkError unstable="false">
-    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java</fileName>
     <specifier>dereference.of.nullable</specifier>
     <message>dereference of possibly-null reference currentClass</message>
     <lineContent>result = currentClass.getDeclaredField(propertyName);</lineContent>

--- a/src/main/java/com/puppycrawl/tools/checkstyle/site/ClassAndPropertiesSettersJavadocScraper.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/site/ClassAndPropertiesSettersJavadocScraper.java
@@ -23,7 +23,6 @@ import java.beans.Introspector;
 import java.util.Collections;
 import java.util.LinkedHashMap;
 import java.util.Map;
-import java.util.Set;
 import java.util.regex.Pattern;
 
 import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
@@ -47,12 +46,6 @@ public class ClassAndPropertiesSettersJavadocScraper extends AbstractJavadocChec
      */
     private static final Map<String, DetailNode> JAVADOC_FOR_MODULE_OR_PROPERTY =
             new LinkedHashMap<>();
-
-    /** List of setter methods that the scraper ignores but shouldn't. */
-    private static final Set<String> SUPERCLASS_SETTERS = Set.of(
-        "AbstractFileSetCheck.setFileExtensions",
-        "AbstractHeaderCheck.setHeader"
-    );
 
     /** Name of the module being scraped. */
     private static String moduleName = "";
@@ -88,13 +81,12 @@ public class ClassAndPropertiesSettersJavadocScraper extends AbstractJavadocChec
         final DetailAST blockCommentAst = getBlockCommentAst();
         if (BlockCommentPosition.isOnMethod(blockCommentAst)) {
             final DetailAST methodDef = getParentAst(blockCommentAst, TokenTypes.METHOD_DEF);
-            if (methodDef != null) {
+            if (methodDef != null
+                    && isSetterMethod(methodDef)
+                    && isMethodOfScrapedModule(methodDef)) {
                 final String methodName = methodDef.findFirstToken(TokenTypes.IDENT).getText();
-                if (isSetterMethod(methodDef) && isMethodOfScrapedModule(methodDef)
-                    || SUPERCLASS_SETTERS.contains(moduleName + "." + methodName)) {
-                    final String propertyName = getPropertyName(methodName);
-                    JAVADOC_FOR_MODULE_OR_PROPERTY.put(propertyName, ast);
-                }
+                final String propertyName = getPropertyName(methodName);
+                JAVADOC_FOR_MODULE_OR_PROPERTY.put(propertyName, ast);
             }
 
         }
@@ -169,34 +161,12 @@ public class ClassAndPropertiesSettersJavadocScraper extends AbstractJavadocChec
     private static boolean isSetterMethod(DetailAST ast) {
         boolean setterMethod = false;
 
-        // Check have a method with exactly 7 children which are all that
-        // is allowed in a proper setter method which does not throw any
-        // exceptions.
-        final int setterGetterMaxChildren = 7;
-        if (ast.getType() == TokenTypes.METHOD_DEF
-                && ast.getChildCount() == setterGetterMaxChildren) {
+        if (ast.getType() == TokenTypes.METHOD_DEF) {
             final DetailAST type = ast.findFirstToken(TokenTypes.TYPE);
             final String name = type.getNextSibling().getText();
             final Pattern setterPattern = Pattern.compile("^set[A-Z].*");
-            final boolean matchesSetterFormat = setterPattern.matcher(name).matches();
 
-            final DetailAST params = ast.findFirstToken(TokenTypes.PARAMETERS);
-            final boolean singleParam = params.getChildCount(TokenTypes.PARAMETER_DEF) == 1;
-
-            if (matchesSetterFormat && singleParam) {
-                // Now verify that the body consists of:
-                // SLIST -> EXPR -> ASSIGN
-                // SEMI
-                // RCURLY
-                final DetailAST slist = ast.findFirstToken(TokenTypes.SLIST);
-
-                // Maximum nodes allowed in a body of setter
-                final int setterBodySize = 3;
-                if (slist != null && slist.getChildCount() == setterBodySize) {
-                    final DetailAST expr = slist.getFirstChild();
-                    setterMethod = expr.getFirstChild().getType() == TokenTypes.ASSIGN;
-                }
-            }
+            setterMethod = setterPattern.matcher(name).matches();
         }
         return setterMethod;
     }

--- a/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java
@@ -452,10 +452,43 @@ public final class SiteUtil {
         final Map<String, DetailNode> unmodifiableJavadocs =
                 ClassAndPropertiesSettersJavadocScraper.getJavadocsForModuleOrProperty();
         final Map<String, DetailNode> javadocs = new LinkedHashMap<>(unmodifiableJavadocs);
+
         properties.forEach(property -> {
-            javadocs.putIfAbsent(property, SUPER_CLASS_PROPERTIES_JAVADOCS.get(property));
+            final DetailNode superClassPropertyJavadoc =
+                    SUPER_CLASS_PROPERTIES_JAVADOCS.get(property);
+            if (superClassPropertyJavadoc != null) {
+                javadocs.putIfAbsent(property, superClassPropertyJavadoc);
+            }
         });
+
+        assertAllPropertySetterJavadocsAreFound(properties, moduleName, javadocs);
+
         return javadocs;
+    }
+
+    /**
+     * Assert that each property has a corresponding setter javadoc that is not null.
+     * 'tokens' and 'javadocTokens' are excluded from this check, because their
+     * description is different from the description of the setter.
+     *
+     * @param properties the properties of the module.
+     * @param moduleName the name of the module.
+     * @param javadocs the javadocs of the properties of the module.
+     * @throws MacroExecutionException if an error occurs during processing.
+     */
+    private static void assertAllPropertySetterJavadocsAreFound(
+            Set<String> properties, String moduleName, Map<String, DetailNode> javadocs)
+            throws MacroExecutionException {
+        for (String property : properties) {
+            final boolean isPropertySetterJavadocFound = javadocs.containsKey(property)
+                       || TOKENS.equals(property) || JAVADOC_TOKENS.equals(property);
+            if (!isPropertySetterJavadocFound) {
+                final String message = String.format(Locale.ROOT,
+                        "%s: Failed to find setter javadoc for property '%s'",
+                        moduleName, property);
+                throw new MacroExecutionException(message);
+            }
+        }
     }
 
     /**

--- a/src/xdocs/checks/coding/matchxpath.xml.template
+++ b/src/xdocs/checks/coding/matchxpath.xml.template
@@ -37,22 +37,10 @@
 
       <subsection name="Properties" id="Properties">
         <div class="wrapper">
-          <table>
-            <tr>
-              <th>name</th>
-              <th>description</th>
-              <th>type</th>
-              <th>default value</th>
-              <th>since</th>
-            </tr>
-            <tr>
-              <td>query</td>
-              <td>Specify Xpath query.</td>
-              <td><a href="../../property_types.html#String">String</a></td>
-              <td><code>""</code></td>
-              <td>8.39</td>
-            </tr>
-          </table>
+          <macro name="properties">
+            <param name="modulePath"
+              value="src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MatchXpathCheck.java"/>
+          </macro>
         </div>
       </subsection>
 

